corehq/motech/repeaters/repeater_generators.py

```diff
@@ -261,6 +270,103 @@ def get_test_payload(self, domain):
 
+@attr.s
+class CaseTypeReferralConfig(object):
+    #if false, listed_properties is a whitelist
+    use_blacklist = attr.ib()
+    listed_properties = attr.ib()
+    constant_properties = attr.ib()
+
+
+class ReferCasePayloadGenerator(BasePayloadGenerator):
+
+    def get_payload(self, repeat_record, payload_doc):
+
+        def _update_case_id(original_case_id, case_id_map):
+            if original_case_id in case_id_map:
+                new_case_id = case_id_map[original_case_id]
+            else:
+                new_case_id = uuid4().hex
+                case_id_map[original_case_id] = new_case_id
+            return new_case_id
+
+        def _update_case_properties_with_blacklist(case, config):
+            for name in config.listed_properties:
+                if name in case.case_json:
+                    del case.case_json[name]
+
+        def _update_case_properties_with_whitelist(case, config):
+            new_json = {}
+            for name in config.listed_properties:
+                if name in case.case_json:
+                    new_json[name] = case.case_json[name]
+            case.case_json = new_json
+
+        def _set_constant_properties(case, config):
+            for name, value in config.constant_properties:
+                case.case_json[name] = value
+
+        case_ids_to_forward = payload_doc.get_case_property('cases_to_forward').split(' ')
+        new_owner = payload_doc.get_case_property('new_owner')
+        cases_to_forward = CaseAccessors(payload_doc.domain).get_cases(case_ids_to_forward)
+        case_ids_to_forward = set(case_ids_to_forward)
+        included_case_types = payload_doc.get_case_property('case_types').split(' ')
+        case_type_configs = {}
+        for case_type in included_case_types:
+            constant_properties = []
+            for key, value in payload_doc.case_json.items():
+                constant_prefix = f'{case_type}_setter_'
+                if key.startswith(constant_prefix):
+                    property_name = key[len(constant_prefix):]
+                    constant_properties.append((property_name, value))
+            whitelist = payload_doc.get_case_property(f'{case_type}_whitelist')
+            blacklist = payload_doc.get_case_property(f'{case_type}_blacklist')
+            if blacklist and whitelist:
+                raise ReferralError(f'both blacklist and whitelist included for {case_type}')
+            if not blacklist and not whitelist:
+                raise ReferralError(f'blacklist or whitelist not included for {case_type}')
+            if blacklist:
+                listed_properties = blacklist
+                use_blacklist = True
+            else:
+                listed_properties = whitelist
+                use_blacklist = False
+            case_type_configs[case_type] = CaseTypeReferralConfig(
+                use_blacklist,
+                listed_properties,
+                constant_properties
+            )
+
+        case_blocks = []
```

Comment:
This section, does the actual case copying. It pulls goes over the cases being moved and changes the properties based on the configuration object created above.

Reply:
good candidate for extracting as a named function